## Group 7
## Big Data Group Project

### Phase 2: Cleansing and Reshaping: Removing Duplicates & Columns with more than 30% null values 

#### Removing Columns with more than 30% null values

Insert libraries into the cell below:

In [1]:
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._
import org.apache.spark.sql._
import org.apache.hadoop.io.LongWritable
import org.apache.hadoop.io.Text
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.mapreduce.lib.input.TextInputFormat

Intitializing Scala interpreter ...

Spark Web UI available at http://95a19fa2a889:4040
SparkContext available as 'sc' (version = 2.4.3, master = local[*], app id = local-1558789912679)
SparkSession available as 'spark'


import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._
import org.apache.spark.sql._
import org.apache.hadoop.io.LongWritable
import org.apache.hadoop.io.Text
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.mapreduce.lib.input.TextInputFormat


Load the data into a scala dataframe:

In [1]:
val df = spark.read.format("com.databricks.spark.csv").option("header", "true").load("m_train_small.csv")

Intitializing Scala interpreter ...

Spark Web UI available at http://95a19fa2a889:4040
SparkContext available as 'sc' (version = 2.4.3, master = local[*], app id = local-1558848888025)
SparkSession available as 'spark'


df: org.apache.spark.sql.DataFrame = [EXT_SOURCE_1: string, EXT_SOURCE_2: string ... 359 more fields]


Count the rows in the dataframe:

In [7]:
df.count()

res3: Long = 307511


Check the percentage of NotNull value of each column:

In [6]:
import org.apache.spark.sql.functions.{col, count, when}

val df4=df.select(df.columns.map(c => (count(c) / count("*")).alias(c)): _*)

import org.apache.spark.sql.functions.{col, count, when}
df4: org.apache.spark.sql.DataFrame = [EXT_SOURCE_1: double, EXT_SOURCE_2: double ... 359 more fields]


In [7]:
df4.show()

+------------------+------------------+------------------+---------------------------------------+----------+----------+------------------+-------------+------------------------------+----------------------+---------------+----------------------------------+----------------------------+-------------------------------+-------------------+-------------------------------+-----------------------------------------+-----------------------------------+----------------------+-------------------------+--------------------------+-----------------+-------------------------------------------+-----------------------------------+-------------+--------------------------+-----------------------------------+-----------------------------+-------------------------+-------------------------------------------+---------------------------+---------------------------------------------------+-----------------------+----------------------------------------+-------------------------------------------+-----------

|      EXT_SOURCE_1|      EXT_SOURCE_2|      EXT_SOURCE_3|client_installments_AMT_PAYMENT_min_sum|DAYS_BIRTH|AMT_CREDIT|       AMT_ANNUITY|DAYS_EMPLOYED|bureau_DAYS_CREDIT_ENDDATE_max|bureau_DAYS_CREDIT_max|DAYS_ID_PUBLISH|bureau_AMT_CREDIT_MAX_OVERDUE_mean|bureau_DAYS_ENDDATE_FACT_max|bureau_AMT_CREDIT_SUM_DEBT_mean|        OWN_CAR_AGE|previous_loans_CNT_PAYMENT_mean|client_cash_CNT_INSTALMENT_FUTURE_min_max|previous_loans_SELLERPLACE_AREA_max|DAYS_LAST_PHONE_CHANGE|bureau_AMT_CREDIT_SUM_max|bureau_AMT_CREDIT_SUM_mean|DAYS_REGISTRATION|client_installments_DAYS_INSTALMENT_max_max|previous_loans_AMT_DOWN_PAYMENT_max|CODE_GENDER_F|REGION_POPULATION_RELATIVE|client_cash_CNT_INSTALMENT_max_mean|bureau_DAYS_CREDIT_UPDATE_max|bureau_AMT_CREDIT_SUM_sum|client_credit_AMT_DRAWINGS_CURRENT_mean_max|YEARS_BEGINEXPLUATATION_AVG|client_installments_NUM_INSTALMENT_VERSION_mean_max|bureau_DAYS_CREDIT_mean|client_installments_AMT_PAYMENT_sum_mean|client_installments_DAYS_INSTALMENT_min_max|previous_lo

TUAL_min_sum|previous_loans_HOUR_APPR_PROCESS_START_min|client_credit_AMT_INST_MIN_REGULARITY_max_min|client_cash_NAME_CONTRACT_STATUS_Signed_count_norm_max|OCCUPATION_TYPE_Laborers|client_cash_CNT_INSTALMENT_min_max|previous_loans_WEEKDAY_APPR_PROCESS_START_THURSDAY_count_norm|ORGANIZATION_TYPE_Construction|client_cash_NAME_CONTRACT_STATUS_Active_count_max|client_cash_SK_DPD_DEF_sum_min|previous_loans_NAME_CONTRACT_TYPE_Cash loans_count_norm|previous_loans_DAYS_FIRST_DRAWING_mean|client_credit_CNT_DRAWINGS_ATM_CURRENT_max_sum|previous_loans_PRODUCT_COMBINATION_Cash X-Sell: middle_count_norm|previous_loans_NAME_CASH_LOAN_PURPOSE_Other_count_norm|client_bureau_balance_STATUS_0_count_norm_sum|client_installments_NUM_INSTALMENT_VERSION_sum_min|previous_loans_CHANNEL_TYPE_Country-wide_count_norm|client_cash_NAME_CONTRACT_STATUS_Signed_count_sum|AMT_REQ_CREDIT_BUREAU_YEAR|bureau_CREDIT_TYPE_Microloan_count|NAME_INCOME_TYPE_State servant|client_credit_AMT_CREDIT_LIMIT_ACTUAL_sum_max|previous

|0.4361892745300168|0.9978537353135335|0.8017469293781361|                     0.9483985938714388|       1.0|       1.0|0.9999609770057006|          1.0|            0.8495370897301234|    0.8568506492450677|            1.0|                0.5979818608114832|          0.7490951543196829|             0.8329165460747746|0.34009189915157506|             0.9451434257636312|                       0.9411695841774765|                 0.9464929709831518|    0.9999967480838083|        0.856847397328876|         0.856847397328876|              1.0|                         0.9483985938714388|                  0.889740529607071|          1.0|                       1.0|                 0.9411695841774765|           0.8568506492450677|       0.8568506492450677|                         0.2826077766323806|         0.5121898078442723|                                 0.9483985938714388|     0.8568506492450677|                      0.9483985938714388|                         0.9483985938714388|           

 0.9464929709831518|                           0.2826077766323806|                          0.29992748226892696|                                     0.9464929709831518|                                    0.9464929709831518|                            0.2826077766323806|                         0.29992748226892696|             1.0|                                 0.9464929709831518|                      0.8568506492450677|                         0.2826077766323806|                                          0.2826077766323806|                     0.29992748226892696|                          0.9464929709831518|                                          0.9464929709831518|                        0.9464929709831518|                                   0.9464929709831518|                            1.0|                                                 0.9464929709831518|            0.9451434257636312|                        0.9464929709831518|                         1.0|                       

Check the count of null value in each column:

In [8]:
import org.apache.spark.sql.functions.{sum, col}
//df.select(df.columns.map(c => sum(col(c).isNull.cast("int")).alias(c)): _*).show
val df3=df.select(df.columns.map(c => sum(col(c).isNull.cast("int")).alias(c)): _*)

import org.apache.spark.sql.functions.{sum, col}
df3: org.apache.spark.sql.DataFrame = [EXT_SOURCE_1: bigint, EXT_SOURCE_2: bigint ... 359 more fields]


In [9]:
df3.show()

+------------+------------+------------+---------------------------------------+----------+----------+-----------+-------------+------------------------------+----------------------+---------------+----------------------------------+----------------------------+-------------------------------+-----------+-------------------------------+-----------------------------------------+-----------------------------------+----------------------+-------------------------+--------------------------+-----------------+-------------------------------------------+-----------------------------------+-------------+--------------------------+-----------------------------------+-----------------------------+-------------------------+-------------------------------------------+---------------------------+---------------------------------------------------+-----------------------+----------------------------------------+-------------------------------------------+--------------------------------------------

|EXT_SOURCE_1|EXT_SOURCE_2|EXT_SOURCE_3|client_installments_AMT_PAYMENT_min_sum|DAYS_BIRTH|AMT_CREDIT|AMT_ANNUITY|DAYS_EMPLOYED|bureau_DAYS_CREDIT_ENDDATE_max|bureau_DAYS_CREDIT_max|DAYS_ID_PUBLISH|bureau_AMT_CREDIT_MAX_OVERDUE_mean|bureau_DAYS_ENDDATE_FACT_max|bureau_AMT_CREDIT_SUM_DEBT_mean|OWN_CAR_AGE|previous_loans_CNT_PAYMENT_mean|client_cash_CNT_INSTALMENT_FUTURE_min_max|previous_loans_SELLERPLACE_AREA_max|DAYS_LAST_PHONE_CHANGE|bureau_AMT_CREDIT_SUM_max|bureau_AMT_CREDIT_SUM_mean|DAYS_REGISTRATION|client_installments_DAYS_INSTALMENT_max_max|previous_loans_AMT_DOWN_PAYMENT_max|CODE_GENDER_F|REGION_POPULATION_RELATIVE|client_cash_CNT_INSTALMENT_max_mean|bureau_DAYS_CREDIT_UPDATE_max|bureau_AMT_CREDIT_SUM_sum|client_credit_AMT_DRAWINGS_CURRENT_mean_max|YEARS_BEGINEXPLUATATION_AVG|client_installments_NUM_INSTALMENT_VERSION_mean_max|bureau_DAYS_CREDIT_mean|client_installments_AMT_PAYMENT_sum_mean|client_installments_DAYS_INSTALMENT_min_max|previous_loans_NAME_CONTRACT_STATUS_Refused_

_max|OCCUPATION_TYPE_Laborers|client_cash_CNT_INSTALMENT_min_max|previous_loans_WEEKDAY_APPR_PROCESS_START_THURSDAY_count_norm|ORGANIZATION_TYPE_Construction|client_cash_NAME_CONTRACT_STATUS_Active_count_max|client_cash_SK_DPD_DEF_sum_min|previous_loans_NAME_CONTRACT_TYPE_Cash loans_count_norm|previous_loans_DAYS_FIRST_DRAWING_mean|client_credit_CNT_DRAWINGS_ATM_CURRENT_max_sum|previous_loans_PRODUCT_COMBINATION_Cash X-Sell: middle_count_norm|previous_loans_NAME_CASH_LOAN_PURPOSE_Other_count_norm|client_bureau_balance_STATUS_0_count_norm_sum|client_installments_NUM_INSTALMENT_VERSION_sum_min|previous_loans_CHANNEL_TYPE_Country-wide_count_norm|client_cash_NAME_CONTRACT_STATUS_Signed_count_sum|AMT_REQ_CREDIT_BUREAU_YEAR|bureau_CREDIT_TYPE_Microloan_count|NAME_INCOME_TYPE_State servant|client_credit_AMT_CREDIT_LIMIT_ACTUAL_sum_max|previous_loans_DAYS_TERMINATION_max|client_credit_AMT_PAYMENT_TOTAL_CURRENT_max_max|FLAG_WORK_PHONE|previous_loans_NAME_GOODS_CATEGORY_Photo / Cinema Equipment_

|      173378|         660|       60965|                                  15868|         0|         0|         12|            0|                         46269|                 44020|              0|                            123625|                       77156|                          51380|     202929|                          16869|                                    18091|                              16454|                     1|                    44021|                     44021|                0|                                      15868|                              33906|            0|                         0|                              18091|                        44020|                    44020|                                     220606|                     150007|                                              15868|                  44020|                                   15868|                                      15868|                                            

              0|                                              16454|                                   44020|                                     220606|                                                      220606|                                  215280|                                       16454|                                                       16454|                                     16454|                                                16454|                              0|                                                              16454|                         16869|                                     16454|                           0|                                  0|                           220606|                                                  18067|                                            16454|                                                    16454|                                            15868|                                   220606|             

Or: Express the count of null value in each column in forms of array:

In [2]:
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types.IntegerType

import org.apache.spark.sql.functions._
import org.apache.spark.sql.types.IntegerType


In [4]:
val col= df.columns
var dfArray=col.map(colmn=>df.select(lit(colmn).as("colName"),sum(when(df(colmn).isNull,1).otherwise(0)).as("missingValues")))
dfArray.tail.foldLeft(dfArray.head)((acc,itr)=>acc.union(itr)).show

2019-05-25 11:42:17,072 WARN  [Thread-4] util.Utils (Logging.scala:logWarning(66)) - Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.debug.maxToStringFields' in SparkEnv.conf.
+--------------------+-------------+
|             colName|missingValues|
+--------------------+-------------+
|        EXT_SOURCE_1|       173378|
|        EXT_SOURCE_2|          660|
|        EXT_SOURCE_3|        60965|
|client_installmen...|        15868|
|          DAYS_BIRTH|            0|
|          AMT_CREDIT|            0|
|         AMT_ANNUITY|           12|
|       DAYS_EMPLOYED|            0|
|bureau_DAYS_CREDI...|        46269|
|bureau_DAYS_CREDI...|        44020|
|     DAYS_ID_PUBLISH|            0|
|bureau_AMT_CREDIT...|       123625|
|bureau_DAYS_ENDDA...|        77156|
|bureau_AMT_CREDIT...|        51380|
|         OWN_CAR_AGE|       202929|
|previous_loans_CN...|        16869|
|client_cash_CNT_I...|        18091|
|previous_loan

col: Array[String] = Array(EXT_SOURCE_1, EXT_SOURCE_2, EXT_SOURCE_3, client_installments_AMT_PAYMENT_min_sum, DAYS_BIRTH, AMT_CREDIT, AMT_ANNUITY, DAYS_EMPLOYED, bureau_DAYS_CREDIT_ENDDATE_max, bureau_DAYS_CREDIT_max, DAYS_ID_PUBLISH, bureau_AMT_CREDIT_MAX_OVERDUE_mean, bureau_DAYS_ENDDATE_FACT_max, bureau_AMT_CREDIT_SUM_DEBT_mean, OWN_CAR_AGE, previous_loans_CNT_PAYMENT_mean, client_cash_CNT_INSTALMENT_FUTURE_min_max, previous_loans_SELLERPLACE_AREA_max, DAYS_LAST_PHONE_CHANGE, bureau_AMT_CREDIT_SUM_max, bureau_AMT_CREDIT_SUM_mean, DAYS_REGISTRATION, client_installments_DAYS_INSTALMENT_max_max, previous_loans_AMT_DOWN_PAYMENT_max, CODE_GENDER_F, REGION_POPULATION_RELATIVE, client_cash_CNT_INSTALMENT_max_mean, bureau_DAYS_CREDIT_UPDATE_max, bureau_AMT_CREDIT_SUM_sum, client_credit_AMT_D...

From the above outcome we can see that there are lots of columns with a big portion of null value.
But if we remove all rows containing 'null', there would be only 800+ rows and 300+columns left, far from enought for
the following steps.

Hence we set a threshold as 92253, which is 30% of the number of total rows. And set a pipeline to select the columns 
in which the null counts is less than 92253, in other words, less than 30% of the column is emply.

In [3]:
val null_thresh = 92253              

val to_keep = df.columns.filter(
    c => df.agg(
        sum(when(df(c).isNull, 1).otherwise(0)).alias(c)
    ).first().getLong(0) <= null_thresh
)

//df.select(to_keep.head, to_keep.tail: _*).show()

2019-05-25 13:16:07,393 WARN  [Thread-4] util.Utils (Logging.scala:logWarning(66)) - Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.debug.maxToStringFields' in SparkEnv.conf.


null_thresh: Int = 92253
to_keep: Array[String] = Array(EXT_SOURCE_2, EXT_SOURCE_3, client_installments_AMT_PAYMENT_min_sum, DAYS_BIRTH, AMT_CREDIT, AMT_ANNUITY, DAYS_EMPLOYED, bureau_DAYS_CREDIT_ENDDATE_max, bureau_DAYS_CREDIT_max, DAYS_ID_PUBLISH, bureau_DAYS_ENDDATE_FACT_max, bureau_AMT_CREDIT_SUM_DEBT_mean, previous_loans_CNT_PAYMENT_mean, client_cash_CNT_INSTALMENT_FUTURE_min_max, previous_loans_SELLERPLACE_AREA_max, DAYS_LAST_PHONE_CHANGE, bureau_AMT_CREDIT_SUM_max, bureau_AMT_CREDIT_SUM_mean, DAYS_REGISTRATION, client_installments_DAYS_INSTALMENT_max_max, previous_loans_AMT_DOWN_PAYMENT_max, CODE_GENDER_F, REGION_POPULATION_RELATIVE, client_cash_CNT_INSTALMENT_max_mean, bureau_DAYS_CREDIT_UPDATE_max, bureau_AMT_CREDIT_SUM_sum, client_installments_NUM_INSTALMENT_VERSION_mean_max, ...

In [6]:
val df5=df.select(to_keep.head, to_keep.tail: _*)

df5: org.apache.spark.sql.DataFrame = [EXT_SOURCE_2: string, EXT_SOURCE_3: string ... 283 more fields]


In [4]:
df5.count()

res2: Long = 307511


We can see that there are 283 columns left and 307511 rows left

Now we can try to drop the null rows and see what we've got:

In [2]:
val df6=df5.na.drop()
df6.count()

2019-05-26 14:02:09,345 WARN  [Thread-4] util.Utils (Logging.scala:logWarning(66)) - Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.debug.maxToStringFields' in SparkEnv.conf.


df6: org.apache.spark.sql.DataFrame = [EXT_SOURCE_2: string, EXT_SOURCE_3: string ... 283 more fields]
res0: Long = 183875


#### Remove duplicates rows:

In [3]:
df6.dropDuplicates("SK_ID_CURR").count()

res1: Long = 183875


We can see that there are no duplicate rows in this dataframe. 
We end up with 183875 rows and 283 columns

Now write our dataframe in a csv file:

In [ ]:
df6.write.option("header", "true").csv("m_train_small_5.csv")